### ISSUE TITLE
6.7: Build cache is missing files

### ISSUE THREAD

**Issue by fawind** [gradle/gradle#14954]
### Expected Behavior

We have a custom Cacheable Task ("MyCacheableTask") that generates configuration files and writes them to multiple output directories. The task is cached in the remote or local Gradle build cache. Inspecting the published build cache for this task on Gradle 6.6.1 or earlier versions, I can see that the build cache for "MyCacheableTask" contains all output directories and generated files and matches the files that got written to the output directories on disk by "MyCacheableTask".

Example cache:
```
# Content of ~/.gradle/caches/build-cache-1/<cache key>
├── METADATA
├── tree-outputDirectories.serviceA.var%2Fconf
│   ├── encrypted-config-value.key
│   ├── service-config.yml
│   ├── [...]
├── tree-outputDirectories.serviceB.var%2Fconf
│   ├── [...]
└── tree-outputDirectories.serviceC.var%2Fconf
    └── [...]
```

### Current Behavior

After upgrading from Gradle 6.6.1 to 6.7, builds of services started failing due to the Gradle build cache of "MyCacheableTask" containing an empty directory without the generated config files. Only one of the output directories is always empty (e.g. `tree-outputDirectories.serviceB.var%2Fconf`, this differs between services). The other cache entries contain the correct generated files.

Looking at the CI history, I could find the build that published the cache. It was the first build after updating to Gradle 6.7. In this build, "MyCacheableTask" finished successfully and wrote the generated config files to the output directories on disk. However, the Gradle build cache published by this job contained the empty output directory without the generated configuration files. Based on that, subsequent builds use the same cache key for "MyCacheableTask" and fail during later tasks because they expect these configuration files to be present.

The empty build cache is only published when running larger parts of Gradle task graph. When only running "MyCacheableTask" individually, the published build cache contains all expected files.

Example cache with missing files:
```
# Content of ~/.gradle/caches/build-cache-1/<cache key>
├── METADATA
├── tree-outputDirectories.serviceA.var%2Fconf
│   ├── encrypted-config-value.key
│   ├── service-config.yml
│   ├── [...]
├── tree-outputDirectories.serviceB.var%2Fconf    # <-- this dir is empty  
└── tree-outputDirectories.serviceC.var%2Fconf
    └── [...]
```

### Context

We have been using "MyCacheable" task for multiple years across our fleet to generate deployment config files based on templates. Caching this task is important as it can take minutes to generate the files. This is currently blocking our roll out of Gradle 6.7.

For registering the output directories in "MyCacheableTask", we are using a provider accepting a `Map<String, File>` (cache identifier mapping to output directory). Looking at logs of this task, the output directories seem to be set correctly:

```java
@CacheableTask
public class MyCacheableTask extends DefaultTask {
   
  // [...]

  public MyCacheableTask() {
    // getOutputDirectories returns a 'Map<String, File>' mapping cache keys to the concrete output directory.
    // E.g. "outputDirectories.serviceA.var/conf" -> ./serviceA/var/conf,
    //         "outputDirectories.serviceB.var/conf" -> ./serviceB/var/conf
    getOutputs().dirs(getProject().provider(this::getOutputDirectories)).withPropertyName("outputDirectories");
  }
  
  // [...]
}
```

"MyCacheableTask" is quite simple, used extensively, and the code has been quite stable during the last years. I would be surprised if this is the cause but happy to provide further info here if necessary!

### Steps to Reproduce 

I have a company internal consistent reproducer where "MyCacheableTask" succeeds with correctly populated output directories but the published build cache contains an empty directory.
Reproducing this was quite hard. I am still not able to consistently reproduce this on my local machine or with a smaller subset of our Gradle task graph. For now, I have an internal CircleCI job running the whole Gradle build task graph that consistently publishes an incorrect build cache.

Unfortunately, I wasn't able to get an open-source reproducer working yet but here are as many infos as possible:

How I reproduced this on my internal repo:
1. Upgrade the Gradle wrapper from 6.6.1 to 6.7.
1. Enable the local build cache and disable the remote build cache (no difference, just for easier debugging)
1. Do a fresh clone of my repo and clear gradle build caches: `rm ~/.gradle/caches/build-cache-1/* && ./gradlew clean`.
1. Run the build task which has "MyCacheableTask" as a task dependency `./gradlew --info -Dorg.gradle.caching.debug=true build`.
1. "MyCacheableTask" and ":build" both succeed. The correct config files are written to the output dirs. The cache of "MyCacheableTask" is published to the local build cache.
1. Extracting the local build cache using `tar -zxvf ~/.gradle/caches/build-cache-1/<cache key>`, I can see that one of the output directories is empty and does not contain the configuration files.
1. Subsequent builds that use the task fail because they use the same cache key and the empty cache directory gets extracted which is missing the configuration files.
1. Running just `./gradlew myCacheableTask` or smaller task subsets of `./gradlew build` or with Gradle 6.6.1 and earlier works as intended and produces a correct build cache containing all files.

Relevant gradle log lines from a build where the build cache contains the empty directory:
```
> Task :myCacheableTask
Appending implementation to build cache key: com.package.MyCacheableTask_Decorated@e3d84a7607324674a538292a4063db66
Appending additional implementation to build cache key: com.package.MyCacheableTask_Decorated@e3d84a7607324674a538292a4063db66
Appending input value fingerprint for 'configurationInputs' to build cache key: c963144a2ed4730695523dc9277dbfa1
Appending input file fingerprints for 'renderDist' to build cache key: 22439d2fada00fda6e4e55ba0a496dfa
Appending input file fingerprints for 'templatedFile' to build cache key: eb3781c76d8f0169a097b61c5678570a
Appending output property name to build cache key: outputDirectories.serviceA.var/conf
Appending output property name to build cache key: outputDirectories.serviceB.var/conf
Appending output property name to build cache key: outputDirectories.serviceC.var/conf
Build cache key for task ':myCacheableTask' is f9351cc4d29068a739d3de8f6cbe5bba
Task ':myCacheableTask' is not up-to-date because:
  No history is available.
Starting process 'command '/home/circleci/project/build/configRenderer/dist/bin/render.sh''.
Successfully started process 'command '/home/circleci/project/build/configRenderer/dist/bin/render.sh''
Stored cache entry for task ':myCacheableTask' with cache key f9351cc4d29068a739d3de8f6cbe5bba
:myCacheableTask (Thread[Execution worker for ':' Thread 2,5,main]) completed. Took 5.706 secs.
```

"MyCacheableTask" details from the build scan timeline:
```
Build cache result | Miss (local), Store (local) |  
-- | -- | --
Cache key | f9351cc4d29068a739d3de8f6cbe5bba |
Cache artifact | 5.21 kB / 17 entries |
Pack | 0.002s |
```

I am expecting 22 entries but 5 files are missing. On Gradle 6.6.1 or when just running the task alone the cache contains all 22 entries. 

### Versions Affected

Using my internal reproducer, I tried bisecting the [nightly releases](https://services.gradle.org/distributions-snapshots/) from `gradle-6.7-20200723004559+0000` (good) to `gradle-6.7-20201019230050+0000` (bad). The first bad snapshot was `gradle-6.7-20200727220037+0000`. `gradle-6.7-20200726220035+0000` was still good ([bisect log](https://gist.github.com/fawind/5ff05aee556fdb61b2ed871684f9a54b)). Looking at the commits in this range, https://github.com/gradle/gradle/pull/13920 might be relevant but just guessing here.

Hope this helps but please take this bisect with a grain of salt. Currently, I can only reproduce consistently on CircleCI and not locally so I might be missing something.

### Your Environment

Gradle 6.7, Gradle Enterprise plugin 3.3.4
Operating system: Linux 4.4.0-1113-aws  
Java runtime: Azul OpenJDK Runtime Environment 15+36
Java VM: Azul OpenJDK 64-Bit Server VM 15+36 (mixed mode, sharing)

We are also on a Gradle enterprise contract and can provide you with internal build scans or other internal data through that channel if necessary.
Happy to iterate on this and provide you with further info if you need anything!

**Comment by big-guy** [comment]
@gradle/execution any ideas what could have changed here?

**Comment by wolfs** [comment]
@fawind I don't know about any change which could have caused this. Can you share a build scan for a failing build and a build afterwards where it is fixed again on the same changes? The build before the failing build would also be interesting. Thank you!

**Comment by wolfs** [comment]
I also wonder if you have some tasks which write to the locations where the output files of `myCacheableTask` are located. Or some tasks which read the files in parallel to `myCacheableTask`, i.e. a missing task dependency.

**Comment by fawind** [comment]
Sorry for the delayed response and thanks for the reply @wolfs, that was totally the right direction!

We were indeed missing a task dependency which resulted in a race condition with these two tasks invalidating FS snapshots when run in parallel. Adding the task dependency fixed it!

Dumping some additional details on what went wrong I discovered while debugging:

1. On "myCacheableTask" task start, Gradle would mark the task as not up-to-date and invalidate existing FS snapshot for all output dirs: `serviceA/var/conf`, `serviceB/var/conf`, ... ([source](https://github.com/gradle/gradle/blob/ccf34e4330db01f4d32863cd0b97ecaec0cb8110/subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java#L203))
2. While "myCacheableTask" is running, "myOtherTask" would do a read access on `serviceA/var`. Because no current snapshot exists, this would create a new snapshot of the currently empty directory ([source](https://github.com/gradle/gradle/blob/ccf34e4330db01f4d32863cd0b97ecaec0cb8110/subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java#L196)).
3. When "myCacheableTask" finishes, Gradle tries to compare snapshots of before and after the task run to determine output files. For serviceB and serviceC (e.g. `serviceB/var/conf`), the last snapshot got properly invalidated in (1.), is not present, and Gradle makes a new snapshot from disk. However for `serviceA/var/conf`, the previous snapshot from (2.) is found and used as the output snapshot without checking the output on disk again ([source](https://github.com/gradle/gradle/blob/ccf34e4330db01f4d32863cd0b97ecaec0cb8110/subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java#L190)).
4. Based on that, Gradle uses the empty snapshot of `serviceA/var/conf` as part of the task output of "myCacheableTask" which results in this directory being empty in the published build cache.

I still believe that some behavior changed here going from 6.6.1 to 6.7 but given that we were missing a task dependency we were relying on undefined behavior so happy to close this issue. Thanks again for pointing us in the right direction!

Maybe as some feedback or feature request: I found this quite hard to debug and with a growing number of custom gradle plugins, I can see a missing a task dependency happening again. From my rudimentary understanding of Gradle, it should have information about task dependencies and which tasks access which parts of the snapshot tree. Based on that, Gradle could warn when two tasks without a dependency access the same FS snapshot. I think there is already a warning when you have overlapping output directories but it looks like this case slipped through.
